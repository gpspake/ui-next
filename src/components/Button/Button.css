.button {
    font-size: 60px;
    background-color: lightcoral;
}
