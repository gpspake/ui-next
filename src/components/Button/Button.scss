:root {
    --button-height--small: 30px;
    --button-height--medium: 36px;
    --button-height--large: 44px;
    --button-height-tight--small: 18px;
    --button-height-tight--medium: 22px;
    --button-height-tight--large: 30px;
    --button-hover-transition: all 0.3s ease-in-out 0s;
}

.base {
    composes: button-reset from '../../styles/helpers.local.css';
    display: inline-block;
    border: 2px solid transparent;
    padding: 0 10px;
    line-height: normal;
    text-decoration: none;
    cursor: pointer;
    font-weight: var(--font-weight-normal);
    text-transform: capitalize;
    transition: var(--button-hover-transition);
}

.base[disabled] {
    cursor: not-allowed;
}

/**
 * Colors
 */
.primary {
    border-color: var(--mid-blue);
    background-color: var(--mid-blue);
    color: var(--mid-blue);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: color(var(--mid-blue) tint(20%));
     background-color: color(var(--mid-blue) tint(20%));
     color: color(var(--mid-blue) tint(20%));
 }

&[disabled] {
     border-color: color(var(--mid-blue) lightness(+30%));
     background-color: color(var(--mid-blue) lightness(+30%));
     color: color(var(--mid-blue) lightness(+30%));
 }
}

.primary.boxyNormal {
    box-shadow: 0 2px 4px color(var(--gray));
}

.primary-hc {
    border-color: var(--dark-blue);
    background-color: var(--dark-blue);
    color: var(--dark-blue);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: color(var(--dark-blue) tint(20%));
     background-color: color(var(--dark-blue) tint(20%));
     color: color(var(--dark-blue) tint(20%));
 }

&[disabled] {
     border-color: color(var(--dark-blue) lightness(+45%));
     background-color: color(var(--dark-blue) lightness(+45%));
     color: color(var(--dark-blue) lightness(+45%));
 }
}

.light {
&.active {
     font-weight: bold;
     background-color: var(--gray);
 }

&:focus:not([disabled]),
&:hover:not([disabled]) {
     background-color: var(--light-gray);
 }

&[disabled] {
     border-color: color(var(--gray) lightness(+40%));
     background-color: color(var(--gray) lightness(+40%));
     color: color(var(--dark-navy) lightness(+40%));
 }
}

.grayscale {
    border-color: color(var(--mid-gray) tint(20%));
    background-color: color(var(--mid-gray) tint(20%));
    color: color(var(--mid-gray) tint(20%));

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: var(--mid-gray);
     background-color: var(--mid-gray);
     color: var(--mid-gray);
 }

&[disabled] {
     border-color: color(var(--mid-gray) lightness(+40%));
     background-color: color(var(--mid-gray) lightness(+40%));
     color: color(var(--mid-gray) lightness(+40%));
 }
}

.grayscale-hc {
    border-color: var(--dark-gray);
    background-color: var(--dark-gray);
    color: var(--dark-gray);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: var(--dark-gray);
     background-color: var(--dark-gray);
     color: var(--dark-gray);
 }

&[disabled] {
     border-color: color(var(--dark-gray) lightness(+40%));
     background-color: color(var(--dark-gray) lightness(+40%));
     color: color(var(--dark-gray) lightness(+40%));
 }
}

.status--positive {
    border-color: var(--status-positive);
    background-color: var(--status-positive);
    color: var(--status-positive);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: color(var(--status-positive) tint(20%));
     background-color: color(var(--status-positive) tint(20%));
     color: color(var(--status-positive) tint(20%));
 }

&[disabled] {
     border-color: color(var(--status-positive) lightness(+20%));
     background-color: color(var(--status-positive) lightness(+20%));
     color: color(var(--status-positive) lightness(+20%));
 }
}

.status--warning {
    border-color: var(--status-warning);
    background-color: var(--status-warning);
    color: var(--status-warning);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: color(var(--status-warning) tint(20%));
     background-color: color(var(--status-warning) tint(20%));
     color: color(var(--status-warning) tint(20%));
 }

&[disabled] {
     border-color: color(var(--status-warning) lightness(+20%));
     background-color: color(var(--status-warning) lightness(+20%));
     color: color(var(--status-warning) lightness(+20%));
 }
}

.status--negative {
    border-color: var(--status-negative);
    background-color: var(--status-negative);
    color: var(--status-negative);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: color(var(--status-negative) tint(20%));
     background-color: color(var(--status-negative) tint(20%));
     color: color(var(--status-negative) tint(20%));
 }

&[disabled] {
     border-color: color(var(--status-negative) lightness(+20%));
     background-color: color(var(--status-negative) lightness(+20%));
     color: color(var(--status-negative) lightness(+20%));
 }
}

.status--negative-hc {
    border-color: var(--status-negative-hc);
    background-color: var(--status-negative-hc);
    color: var(--status-negative-hc);

&:focus:not([disabled]),
&:hover:not([disabled]) {
     border-color: color(var(--status-negative-hc) tint(20%));
     background-color: color(var(--status-negative-hc) tint(20%));
     color: color(var(--status-negative-hc) tint(20%));
 }

&[disabled] {
     border-color: color(var(--status-negative-hc) lightness(+20%));
     background-color: color(var(--status-negative-hc) lightness(+20%));
     color: color(var(--status-negative-hc) lightness(+20%));
 }
}

/**
 * Themes
 */
.normal,
.boxyNormal {
    color: var(--white) !important;
}

.boxyLight,
.boxyOutline {
    color: var(--dark-navy);
    background-color: var(--dark-white);
    border: 1px solid var(--light-gray);
}

.boxyOutline {
    border-color: var(--mid-blue);
    color: var(--mid-blue);
}

.outline,
.boxyOutline {
    background-color: transparent !important;
}

.contentOnly {
    background-color: transparent !important;
    border-color: transparent !important;
    padding: 0;
    min-width: auto;
}

.boxyNormal.small,
.boxyLight.small,
.boxyOutline.small {
    border-radius: calc(var(--button-height--small) / 12);
}

.boxyNormal.medium,
.boxyLight.medium,
.boxyOutline.medium {
    border-radius: calc(var(--button-height--medium) / 12);
}

.boxyNormal.large,
.boxyLight.large,
.boxyOutline.large {
    border-radius: calc(var(--button-height--large) / 12);
}

.boxyTight {
    color: var(--white) !important;

&.small {
     height: var(--button-height-tight--small);
     min-width: auto;
     border-radius: 3px;
 }

&.medium {
     height: var(--button-height-tight--medium);
     min-width: auto;
     border-radius: 3px;
 }

&.large {
     height: var(--button-height-tight--large);
     min-width: auto;
     border-radius: 3px;
 }

span {
    line-height: normal;
}
}

/**
 * Sizes
 */
.small {
    height: var(--button-height--small);
    min-width: auto;
    border-radius: calc(var(--button-height--small) / 2);
    font-size: 0.75em;
}

.medium {
    height: var(--button-height--medium);
    min-width: 100px;
    border-radius: calc(var(--button-height--medium) / 2);
    font-size: 0.9em;
}

.large {
    height: var(--button-height--large);
    min-width: 200px;
    border-radius: calc(var(--button-height--large) / 2);
    font-size: 1.25em;
}

/* Button display modifiers */
.block {
    display: block;
    width: 100%;
}

/* Button children styles */
.inner {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    white-space: nowrap;

& > * {
      color: currentColor;
      text-decoration: none;
      border-bottom: none;
  }
}

.spinnerBase {
    position: absolute;
    height: 26px !important;
}

.hidden {
    visibility: hidden;
    height: 0;
}

.icon {
    margin-right: 0.4em;
    flex: 0 0 auto;
}

.iconOnly .icon {
    margin: 0;
}

.text {
    font-family: var(--sans);
    text-align: center;
    flex: 1 0 auto;
}

.child:hover,
.base:hover > * {
    color: currentColor;
}
