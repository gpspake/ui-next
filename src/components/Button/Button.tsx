import React from 'react';
import classNames from 'classnames';
import styles from './Button.scss';
import Icon from '../Icon';
import Spinner from '../Spinner';
import icons from '../Icon/icon-config';

type ButtonProps = {
  /**
   * title attribute
   */
  title: string,
  /**
   * Display the button as a block element
   */
  block?: boolean,
  /**
   * String or element to be displayed inside button
   */
  children?: React.ReactNode,
  /**
   * ClassName will insert styles to override defaults
   */
  className?: string,
  /**
   * Pass custom icon classname into the button to icon
   */
  iconClassName?: string,
  /**
   * Specify a thematic color
   */
  color?: 'primary' |
    'primary-hc' |
    'secondary' |
    'grayscale' |
    'grayscale-hc' |
    'light' |
    'status--positive' |
    'status--warning' |
    'status--negative' |
    'status--positive-hc' |
    'status--warning-hc' |
    'status--negative-hc',
  /**
   * HTML disabled state, non-interactive
   */
  disabled?: boolean,
  /**
   * Show spinner instead of children
   */
  spinner?: boolean,
  /**
   * Add an icon next to the text
   */
  icon?: keyof typeof icons,
  /**
   * A function that is called on button click
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>,
  /**
   * an accessibility label for the button
   */
  ariaLabel?: string,
  /**
   * Choose a size for the button
   */
  size?: 'small' | 'medium' | 'large',

  theme?: 'normal' |
  'outline' |
  'contentOnly' |
  'boxyNormal' |
  'boxyLight' |
  'boxyOutline' |
  'boxyTight'
  /**
   * HTML button type
   */
  type?: 'button' | 'submit'
}

/**
 * Used to attach click events, create links, trigger actions
 */
const Button = (props: ButtonProps) => {

  const {
    title,
    iconClassName = '',
    className = '',
    block = false,
    children = null,
    color = 'primary',
    disabled = false,
    spinner = false,
    icon = null,
    onClick = () => {},
    ariaLabel,
    size = 'medium',
    theme = 'normal',
    type = 'button'
  } = props



  const childComponent =
    typeof children === 'string' || typeof children === 'number' ? (
      <span className={styles.text}>{children}</span>
    ) : (
      children
    );

  return (
    <button
      title={title}
      className={classNames(
        styles.base,
        styles[theme],
        styles[size],
        styles[color],
        block && styles.block,
        className,
        children ? null : styles.iconOnly
      )}
      onClick={onClick}
      disabled={disabled || spinner}
      type={type}
      aria-label={ariaLabel}
    >
      {spinner && (
        <div className={classNames([styles.spinner, styles.inner])}>
          <Spinner
            className={styles.spinnerBase}
            size="small"
            color="currentColor"
          />
        </div>
      )}
      <div
        className={classNames([styles.inner, spinner && styles.hidden])}
      >
        {icon && (
          <Icon
            icon={icon}
            className={classNames(styles.icon, iconClassName)}
          />
        )}
        {childComponent}
      </div>
    </button>
  );
};

Button.displayName = 'UIButton';

export default Button;
