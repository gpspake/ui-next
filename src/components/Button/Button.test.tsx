import React from 'react';
import renderer from 'react-test-renderer';
import Button from './Button';

it('matches snapshot', () => {
  const component = renderer.create(
    <Button title={"my-button"}/>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
