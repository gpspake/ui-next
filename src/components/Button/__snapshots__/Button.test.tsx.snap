// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  className="base normal medium primary iconOnly"
  disabled={false}
  onClick={[Function]}
  title="my-button"
  type="button"
>
  <div
    className="inner"
  />
</button>
`;
